### PR TITLE
Adds Vault Verb

### LINKED ISSUE #1883: Table Vault Verb

### PR DESCRIPTION
**Pull request by exp111**
Fixes #1883

### CODE REVIEW

File: Content.Server/GameObjects/Components/Movement/ClimbableComponent.cs [hunk 1]
```diff
@@ -229,5 +263,27 @@ private void PopupMessageOtherClientsInRange(IEntity source, string message, int
                 source.PopupMessage(viewer.AttachedEntity, message);
             }
         }
+
+        /// <summary>
+        ///     Allows to vault a object with the ClimbableComponent through right click
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Movement/ClimbableComponent.cs:268]: Allows **you** to vault **an** object

File: Content.Server/GameObjects/Components/Movement/ClimbableComponent.cs [hunk 2]
```diff
@@ -60,68 +61,101 @@ public override void ExposeData(ObjectSerializer serializer)
...
+            string reason;
+            bool canVault;
+
+            if (eventArgs.User == eventArgs.Dropped)
+                canVault = CanVault(eventArgs.User, eventArgs.Target, out reason);
+            else
+                canVault = CanVault(eventArgs.User, eventArgs.Dropped, eventArgs.Target, out reason);
 
+            if (!canVault)
+                _notifyManager.PopupMessage(eventArgs.User, eventArgs.User, reason);
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Movement/ClimbableComponent.cs:73]: Owner.PopupMessage?

File: Content.Server/GameObjects/Components/Movement/ClimbableComponent.cs [hunk 3]
```diff
@@ -60,68 +61,101 @@ public override void ExposeData(ObjectSerializer serializer)
...
-                    return false;
-                }
+            if (!interaction.InRangeUnobstructed(userPosition, climbablePosition, _range, predicate: Ignored) ||
+                !interaction.InRangeUnobstructed(userPosition, otherUserPosition, _range, predicate: Ignored))
+            {
+                reason = Loc.GetString("You can't reach there!");
+                return false;
             }
 
+            reason = "";
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Movement/ClimbableComponent.cs:158]: Above uses string.empty but this uses "". Also should this be nullable?

File: Content.Server/GameObjects/Components/Movement/ClimbableComponent.cs [hunk 4]
```diff
@@ -60,68 +61,101 @@ public override void ExposeData(ObjectSerializer serializer)
...
-            else // user is dragging some other entity onto a climbable
+        /// <summary>
+        /// Checks if the user can vault the dragged entity onto the the target
+        /// </summary>
+        /// <param name="user">The user that wants to vault the entity</param>
+        /// <param name="dragged">The entity that is being vaulted</param>
+        /// <param name="target">The object that is being vaulted onto</param>
+        /// <param name="reason">The reason why it cant be dropped</param>
+        /// <returns></returns>
+        private bool CanVault(IEntity user, IEntity dragged, IEntity target, out string reason)
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Movement/ClimbableComponent.cs:131]: This has duplicate checks with the other CanVault method, is it possible to combine these a bit more?

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Server/GameObjects/Components/Movement/ClimbableComponent.cs (hunk 1 above)
Comment (RemieRichards) [reply Content.Server/GameObjects/Components/Movement/ClimbableComponent.cs:268]: esl exp 😢

File: Content.Server/GameObjects/Components/Movement/ClimbableComponent.cs (hunk 4 above)
Comment (exp111) [reply Content.Server/GameObjects/Components/Movement/ClimbableComponent.cs:129]: Mhh you I make a pre function that will take dragged as null and then check the first two things, but I don't think that's better

Comment (metalgearsloth) [reply Content.Server/GameObjects/Components/Movement/ClimbableComponent.cs:129]: Yeah I'm not sure what it should be; you could try changing the internals so they both call a 3rd function with the common functionality, but you could just out of scope it for the pr and leave as is I guess.

**Comment by metalgearsloth** [comment]
Once conflict's done LGTM if you want to leave the duplicate code for now.

**Review by DrSmugleaf — APPROVED** [review]

**Pull request merged**